From 283d1fdf0169c3b6cfa1400846391b42050fa6ea Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Wed, 16 Dec 2015 14:12:57 -0500
Subject: [PATCH 2/2] rtl8139: Do not consume the packet during overflow in
 standard mode.

RH-Author: Vlad Yasevich <vyasevic@redhat.com>
Message-id: <1450275177-24818-3-git-send-email-vyasevic@redhat.com>
Patchwork-id: 68630
O-Subject: [RHEL6.8 qemu-kvm PATCH 2/2] rtl8139: Do not consume the packet during overflow in standard mode.
Bugzilla: 1262866
RH-Acked-by: Xiao Wang <jasowang@redhat.com>
RH-Acked-by: Stefan Hajnoczi <stefanha@redhat.com>
RH-Acked-by: Thomas Huth <thuth@redhat.com>
RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>

When operation in standard mode, we currently return the size
of packet during buffer overflow.  This consumes the overflow
packet.  Return 0 instead so we can re-process the overflow packet
when we have room.

This fixes issues with lost/dropped fragments of large messages.

Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Reviewed-by: Jason Wang <jasowang@redhat.com>
Message-id: 1441121206-6997-3-git-send-email-vyasevic@redhat.com
Signed-off-by: Stefan Hajnoczi <stefanha@redhat.com>
(cherry picked from commit 26c4e7ca72d970d120f0f51244bc8d37458512a0)
Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Jeff E. Nelson <jen@redhat.com>

Conflicts:
	hw/net/rtl8139.c
        -File is actually hw/rtl8139.  Changed by hand.
---
 hw/rtl8139.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/rtl8139.c b/hw/rtl8139.c
index 29e186d..7454683 100644
--- a/hw/rtl8139.c
+++ b/hw/rtl8139.c
@@ -1152,7 +1152,7 @@ static ssize_t rtl8139_do_receive(VLANClientState *nc, const uint8_t *buf, size_
             s->IntrStatus |= RxOverflow;
             ++s->RxMissed;
             rtl8139_update_irq(s);
-            return size_;
+            return 0;
         }
 
         packet_header |= RxStatusOK;
-- 
2.1.0

